mlir/lib/Dialect/Affine/Analysis/Utils.cpp

```diff
@@ -218,12 +233,14 @@ bool MemRefDependenceGraph::init() {
     unsigned n = memrefAndList.second.size();
     for (unsigned i = 0; i < n; ++i) {
       unsigned srcId = memrefAndList.second[i];
-      bool srcHasStore =
-          getNode(srcId)->getStoreOpCount(memrefAndList.first) > 0;
+      Node *srcNode = getNode(srcId);
+      bool srcHasStore = (!srcNode->memEffects.empty()) ||
+                         (srcNode->getStoreOpCount(memrefAndList.first) > 0);
       for (unsigned j = i + 1; j < n; ++j) {
         unsigned dstId = memrefAndList.second[j];
-        bool dstHasStore =
-            getNode(dstId)->getStoreOpCount(memrefAndList.first) > 0;
+        Node *dstNode = getNode(dstId);
+        bool dstHasStore = (!dstNode->memEffects.empty()) ||
```

Comment:
Variable name will be inaccurate now. `dstMayHaveStore`.

Moreover, you could have only kept write memory effects here.